File: backend/src/main/java/com/inventory/backend/repository/PrinterInstallationRepository.java
package com.inventory.backend.repository;

import com.inventory.backend.entity.PrinterInstallation;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.Optional;

public interface PrinterInstallationRepository extends JpaRepository<PrinterInstallation, Long> {

    List<PrinterInstallation> findByCartridgeId(Long cartridgeId);

    Optional<PrinterInstallation> findByCartridgeIdAndPrinterSlotId(Long cartridgeId, Long printerSlotId);

    Optional<PrinterInstallation> findFirstByPrinterSlotIdAndQuantityGreaterThan(Long printerSlotId, Integer quantity);

    long countByPrinterSlotId(Long printerSlotId);
}

File: backend/src/main/kotlin/com/inventory/backend/repository/PrinterInstallationRepository.kt
package com.inventory.backend.repository

import com.inventory.backend.entity.PrinterInstallation
import org.springframework.data.jpa.repository.JpaRepository
import java.util.Optional

interface PrinterInstallationRepository : JpaRepository<PrinterInstallation, Long> {
    fun findByCartridgeId(cartridgeId: Long): List<PrinterInstallation>
    fun findByCartridgeIdAndPrinterSlotId(cartridgeId: Long, printerSlotId: Long): Optional<PrinterInstallation>
    fun findFirstByPrinterSlotIdAndQuantityGreaterThan(printerSlotId: Long, quantity: Int): Optional<PrinterInstallation>
    fun countByPrinterSlotId(printerSlotId: Long): Long
}
